metric_learn/scml.py

```diff
@@ -0,0 +1,596 @@
+    # np.array (2*lenT,2)
+    T_pairs_sorted = np.sort(np.vstack((triplets[:, [0, 1]],
+                                        triplets[:, [0, 2]])),
+                             kind='stable')
+    # calculate all unique pairs and their indices
+    uniqPairs = np.unique(T_pairs_sorted, axis=0)
+
+    if self.n_basis is None:
+      # TODO: Get a good default n_basis directive
+      n_basis = uniqPairs.shape[0]
+      warnings.warn('The number of basis will be set to n_basis= %d' % n_basis)
+
+    elif isinstance(self.n_basis, int):
+      n_basis = self.n_basis
+    else:
+      raise ValueError("n_basis should be an integer, instead it is of type %s"
+                       % type(self.n_basis))
+
+    if n_basis > uniqPairs.shape[0]:
+      n_basis = uniqPairs.shape[0]
+      warnings.warn("The selected number of basis is greater than the number "
+                    "of points, only n_basis = %d will be generated" %
+                    n_basis)
+
+    uniqPairs = X[uniqPairs]
+
+    rng = check_random_state(self.random_state)
+
+    # Select n_basis
+    selected_pairs = uniqPairs[rng.choice(uniqPairs.shape[0],
+                               size=n_basis, replace=False), :, :]
+
+    basis = selected_pairs[:, 0]-selected_pairs[:, 1]
+
+    return basis, n_basis
+
+
+class SCML_global(_BaseSCML_global, _TripletsClassifierMixin):
+  """Sparse Compositional Metric Learning (SCML)
+
+  `SCML` builds a metric as the sparse positive combination of a set of locally
+  discriminative rank-one PSD basis. This allows an optimization scheme with
+  only `K` parameters, that can be yielded with an efficient stochastic
+  composite optimization over a set of triplets constraints. Each triplet is
+  constructed as a relative distance comparison with respect to the first
+  element so that the second element is closer than the last.
+  Read more in the :ref:`User Guide <scml>`.
+  Parameters
+  ----------
+  beta: float (default=1e-5)
+      L1 regularization parameter.
+  basis : None, string or numpy array, optional (default=None)
+       Prior to set for the metric. Possible options are
+       '', and a numpy array of shape (n_basis, n_features). If
+        None an error will be raised as the basis set is esential
+        to SCML.
+       numpy array
+           A matrix of shape (n_basis, n_features), that will be used as
+           the basis set for the metric construction.
+  n_basis : int, optional
+      Number of basis to be yielded. In case it is not set it will be set based
+      on the basis numpy array. If an string option is pased to basis an error
+      wild be raised as this value will be needed.
+  gamma: float (default = 5e-3)
+      Learning rate
+  max_iter : int (default = 100000)
+      Number of iterations for the algorithm
+  output_iter : int (default = 5000)
+      Number of iterations to check current weights performance and output this
+      information in case verbose is True.
```

Comment:
not sure we should uncover this. maybe we could fix it, or choose it based on `max_iter`

Reply:
I uncovered it based on @terrytangyuan suggestion. A fixed value based on `max_iter` sounds good though.

Reply:
We do not expose this for other algorithms: typically the verbose mode outputs each iteration. Of course these are batch algorithms - for stochastic algorithms with low per-iteration complexity, it is a waste of resources to print at each iteration. On the other hand, the purpose of verbose mode is to check what is happening, so we could consider printing every iteration, or say every 10 iter if the computational burden is too high.

Reply:
I agree that we should try to be consistent with other algorithms. Although I think the decision is not so trivial as on every output iteration the objective is computed with all the constraints, so there is a high computational cost compared to each normal iteration. Also, on every output iteration the best weights are updated, so reducing this value could lead to overfitting as more instances (with possible overfitting) are tested.

Maybe a reasonable default value in relation to `max_iter` sounds good?

Reply:
Good point, you also use this to keep the best weight, not only for verbose purposes.
Maybe it is good to expose it then. As we may also use it to adapt the learning rate automatically